vortex-buffer/src/lib.rs

```diff
@@ -2,99 +2,86 @@ use arrow_buffer::Buffer as ArrowBuffer;
 use vortex_dtype::{match_each_native_ptype, NativePType};
 
 #[derive(Debug, Clone)]
-pub enum Buffer<'a> {
-    Owned(ArrowBuffer),
-    View(&'a [u8]),
+pub enum Buffer {
+    // TODO(ngates): we could add Aligned(Arc<AVec>) from aligned-vec package
+    Arrow(ArrowBuffer),
+    Bytes(bytes::Bytes),
 }
 
-pub type OwnedBuffer = Buffer<'static>;
+unsafe impl Send for Buffer {}
+unsafe impl Sync for Buffer {}
 
-impl Buffer<'_> {
+impl Buffer {
     pub fn len(&self) -> usize {
         match self {
-            Buffer::Owned(buffer) => buffer.len(),
-            Buffer::View(slice) => slice.len(),
+            Buffer::Arrow(b) => b.len(),
+            Buffer::Bytes(b) => b.len(),
         }
     }
 
     pub fn is_empty(&self) -> bool {
-        self.len() == 0
-    }
-
-    pub fn as_slice(&self) -> &[u8] {
         match self {
-            Buffer::Owned(buffer) => buffer.as_slice(),
-            Buffer::View(slice) => slice,
+            Buffer::Arrow(b) => b.is_empty(),
+            Buffer::Bytes(b) => b.is_empty(),
         }
     }
 
     pub fn typed_data<T: NativePType>(&self) -> &[T] {
         match self {
-            Buffer::Owned(buffer) => unsafe {
+            Buffer::Arrow(buffer) => unsafe {
                 match_each_native_ptype!(T::PTYPE, |$T| {
                     std::mem::transmute(buffer.typed_data::<$T>())
                 })
             },
-            Buffer::View(slice) => {
+            Buffer::Bytes(bytes) => {
                 // From ArrowBuffer::typed_data
-                let (prefix, offsets, suffix) = unsafe { slice.align_to::<T>() };
+                let (prefix, offsets, suffix) = unsafe { bytes.align_to::<T>() };
                 assert!(prefix.is_empty() && suffix.is_empty());
                 offsets
             }
         }
     }
 
-    pub fn to_static(&self) -> OwnedBuffer {
-        match self {
-            Buffer::Owned(d) => Buffer::Owned(d.clone()),
-            Buffer::View(_) => Buffer::Owned(self.into()),
-        }
-    }
-}
-
-impl<'a> Buffer<'a> {
-    pub fn into_vec<T: NativePType>(self) -> Result<Vec<T>, Buffer<'a>> {
+    pub fn into_vec<T: NativePType>(self) -> Result<Vec<T>, Buffer> {
         match self {
-            Buffer::Owned(buffer) => match_each_native_ptype!(T::PTYPE, |$T| {
+            Buffer::Arrow(buffer) => match_each_native_ptype!(T::PTYPE, |$T| {
                 buffer
                     .into_vec()
                     .map(|vec| unsafe { std::mem::transmute::<Vec<$T>, Vec<T>>(vec) })
-                    .map_err(Buffer::Owned)
+                    .map_err(Buffer::Arrow)
             }),
-            Buffer::View(_) => Err(self),
+            // Cannot always convert bytes into a mutable vec
+            Buffer::Bytes(_) => Err(self),
         }
     }
 }
 
-impl From<ArrowBuffer> for OwnedBuffer {
-    fn from(value: ArrowBuffer) -> Self {
-        Buffer::Owned(value)
+impl AsRef<[u8]> for Buffer {
```

Comment:
The pattern I have seen for this is implementing Deref instead of AsRef. I think AsRef works here but seems not very idiomatic

Reply:
Bytes implements both, Arrow implements deref. I think the difference is subtle, arguably we might want both.

https://github.com/tokio-rs/bytes/blob/86694b05649c0c1666044b2ba5c386c2328aac18/src/bytes.rs#L633-L647

https://docs.rs/arrow-buffer/51.0.0/src/arrow_buffer/buffer/immutable.rs.html#359-365

Reply:
yeah, I would agree that you want both. I noticed this in Arrow and wanted to more so to make sure it was intentional